### ISSUE TITLE
Move mach into a submodule?

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#18343]
For 3rd party projects embedding Servo, on Linux and Mac, `cargo build` is enough to build servo, but mach is necessary on Windows. First to download the necessary dependencies, and then to set the proper env variables.

At the moment, we copy `mach`, `mach.bat` and `/python/` in the embedder code.

From the embedder perspective, it's not great to have to copy and tweak the python scripts.

Would there be a way to "detach" mach from Servo and make it available as a 3rd party tool?

**Comment by SimonSapin** [comment]
How feasible is to make mach unnecessary in the first place? What are these dependencies and environment variables? Could they be handled by Cargo build scripts?

**Comment by paulrouget** [comment]
> Could they be handled by Cargo build scripts?

I believe so. And that would be awesome.

This is mostly needed for Windows.

What is necessary (but probably not enough, still investigating), is:
- downloading the dependencies (mach bootstrap)
- set env variables

A build script in the embedder can not set the variable for dependencies, so that annoying.

**Comment by SimonSapin** [comment]
Rather than trying to solve "set arbitrary env variables from the top-level build system", could we look at what individual variables are currently needed and what we can do about each? Presumably crates that don’t Just Work probably compile code in some non-Rust language with their own build script, which could be controlled by Cargo "features" and be responsible for setting env variables for use by that library’s build system.

**Comment by paulrouget** [comment]
Yes, we could do that.

The python scripts do a lot more than just setting the variables. But I don't see why we couldn't migrate all the features, step by step, into build scripts.

**Comment by shana** [comment]
I needed to build servo-embed-example on windows and the mach dependency was annoying me, so I dug down into what is actually needed in order to build, and https://github.com/shana/servo/commit/afbd29ec314391cc307c376ae329630b0857561d is the result.

It turns out there's only around 5 environment variables needed plus a few path adjustments, all very doable. I have a separate powershell script to download the dependencies, and the main bootstrap.bat script just sets up the command line with the required environment, ready for a `cargo build` invocation.

I didn't do a PR because I know you'd rather have cargo scripts to do this, but I don't know exactly how to set this stuff up in cargo in order for dependencies to pick up the proper environment variables (I wouldn't mind fixing that if someone could point me to docs that show how that would work... but cargo doesn't have postbuild steps, so I'm not sure if everything is possible just with cargo).

I hope having a clean list of the requirements is useful, at least :P

**Comment by SimonSapin** [comment]
Thanks for writing this up @shana! It looks like there is:

* Downloading and installing rustup.rs, but IMO that’s sort of expected for Rust programs.
* After the build, copying some OpenSSL `.dll` files where Servo can find them at runtime. Today’s Cargo can definitely not help there. (There’s no post-build scripts.)
* Several things use Python, so it needs to be in `$PATH`.
* The `script` crate apparently uses CMake and Ninja (so they need to be in `$PATH`), I assume for the code generation from WebIDL. Maybe that could be rewritten in Python or Rust?
* SpiderMonkey / mozjs’s build system apparently requires a number of command-line programs like `ls` and `perl` to be available in `$PATH`. `moztools-0.0.1-5.zip` contains those, pre-compiled for Windows. I don’t know what’s a good way to bootstrap that. The compressed ZIP is 24 MB, maybe too big to ship on crates.io. I wonder what Firefox does there.
* The build script of the `openssl-sys` crate relies on some environment variables to know where to find binaries and header files for OpenSSL. Unfortunately that’s some something we can drive from the build script of another crate. If we want the “download from `servo-deps.s3.amazonaws.com`” scenario to “just work”, `openssl-sys`’s own build script would need to know about it. And forking the crate is not without downsides.

  `openssl-sys` also tries to use vcpkg, maybe that’s a viable alternative but I’m not familiar with it.

So it’s definitely more complex than I hoped in my previous “let’s just write some Cargo build script” comment :)

**Comment by shana** [comment]
As mentioned in some other issue, cargo is a package manager, not a build system (I would argue that if it has a build command then it kinda is a build system but ¯\_(ツ)_/¯), so cargo is probably not the best way to go to drive the whole thing. It would be nice if the build bits in cargo got moved into a separate project that would actually consider itself a build system, but until (or if) that happens, some pre and post build scripts probably need to happen.

I can run the build right now just with path and environment variables set (and doing that postbuild step once), so I think `mach build` is not doing anything special (unless I missed something, anyone know different?).  I haven't looked at all the code for all platforms that the python scripts do for bootstrapping and running the build, but windows is not the only platform that requires environment variables and things in the path, and it's likely the most complicated.

This means that the whole `mach build` thing can probably be replaced with 2 shell scripts, an environment boostrap script and a postbuild script, in 2 variants - bat for windows and sh for the rest. A third script for loading the environment, driving cargo commands, and running any post-command scripts needed can be optionally added as well, for ease of use and isolation testing. Having a script that sets up the shell environment properly would also be very useful for doing things like building apps that depend on servo itself. Does this make sense?

**Comment by JoshMcguigan** [comment]
I was working on a port of mach to Rust when I found this issue. After reading this thread, its not clear to me if a port makes sense (vs finding a way to remove the need for mach entirely), but I'd appreciate any feedback.

In particular, if there is interest in this type of port I'd be happy to continue work on it.

https://github.com/JoshMcguigan/mars

**Comment by raptortech-js** [comment]
After three years, do we still want to get rid of mach?

**Comment by jdm** [comment]
Personally, I only see a benefit to removing mach and giving us a cargo-compatible build. I don't see a benefit to replacing it with some other driver.

**Comment by Eragonfr** [comment]
What features are needed for a basic mach replacement in cargo ?

**Comment by sagudev** [comment]
The topic of this issue is to stop using python but we need python anyway at least for mozjs (SM build system uses python) and wpt test runner, so it is simply more conveniently to keep using python based build system/command runner.

But we should definitely liberate servo of `mach build` and `mach cargo-*` commands.

**Comment by mrobinson** [comment]
Yes, ideally where Python is still necessary, cargo will call the Python code. There are some tricky issues. Just a couple offhand:

 1. Cargo doesn't have support for post-build commands so copying built DLLs around can only happen in build scripts (which must be set up to rerun when those DLLs change).
 2. You can [override environment variables in the cargo configuration](https://doc.rust-lang.org/cargo/reference/config.html#env), but these can only replace existing environment variables. They cannot *extend* those in the environment. This makes the options for updating the `PATH` environment variable pretty tricky.

**Comment by Manishearth** [comment]
>  so it is simply more conveniently to keep using python based build system/command runner.

Yep, we should keep `mach`, but the main thing is that a normal developer should be able to primarily work with `cargo build` and `cargo run`, and they only need to use mach if:

 - They want to install prerequisites
 - They want to deploy to other platforms like Android
 - They want to run special kinds of tests

**Comment by okias** [comment]
As a play with Flatpak builds, I'm would be happy if running `cargo` would be enough, CCing myself here :)
